qiskit/qasm3/ast.py

```diff
@@ -194,16 +165,95 @@ def __init__(
         self.end = end
 
 
+class Identifier(Expression):
+    def __init__(self, string):
+        self.string = string
+
+
 class SubscriptedIdentifier(Identifier):
-    """
-    An identifier with subscripted access.
-    """
+    """An identifier with subscripted access."""
 
-    def __init__(self, identifier: Identifier, subscript: Union[Range, Expression]):
-        self.identifier = identifier
+    def __init__(self, string: str, subscript: Union[Range, Expression]):
+        super().__init__(string)
         self.subscript = subscript
 
 
+class Constant(Expression, enum.Enum):
+    """A constant value defined by the QASM 3 spec."""
+
+    PI = enum.auto()
+    EULER = enum.auto()
+    TAU = enum.auto()
+
+
+class IntegerLiteral(Expression):
```

Comment:
Renaming these classes with correct names is a great idea. There seem to be a number of these improvements that are unrelated to the  PR. One might argue for doing this in a separate PR. But in the interest of development efficiency, I think it's ok to include in this PR.

Reply:
In some senses there's not a lot of renaming going on here - most of the classes I removed weren't actually used, and I was putting in new nodes for the actual `Expr` tree.  All previous `ParameterExpression` values were being handled by the base `ast.Expression` node, which I _effectively_ renamed to `ast.StringifyAndPray` to make it clearer what was actually going on, and so I had a more sane base `ast.Expression` node to work from.

I think `IntegerLiteral` might be the only one that logically existed in the old form that still exists now but with a different name.

Fwiw, I think this AST module will be completely gone by the end of the year; once we have our custom importer, we can use the AST from that, rather than needed to maintain a parallel AST and formatter in Terra.